Add script that generates input files for TOPAS (#25)
SebastiaanBreedveld/TROTS: TOPAS_simulation/README.md

# TOPAS-DICOM Workflow

## Previous requirements

> These scripts are intended to run once the writeTROTStoDICOM.py file has been fully compiled

Reviewer: Mas bien executed, no compiled

> These scripts are intended to run once the writeTROTStoDICOM.py file has been fully compiled

Author: ```suggestion
These scripts are intended to run once the writeTROTStoDICOM.py file has been fully executed.
```

> Estos scripts están pensados para ejecutarse una vez se haya compilado al completo write TROTStoDICOM.py

Reviewer: Thanks. Please translate

Author: oops sorry ajjsaja

> Estos scripts están pensados para ejecutarse una vez se haya compilado al completo write TROTStoDICOM.py

Author: ```suggestion
These scripts are intended to run once the writeTROTStoDICOM.py file has been fully compiled
```

> Ese script ha de haberse compilado con los siguientes argumentos:

Author: ```suggestion
That script must have been compiled with the following arguments:
```

```python
--TreatmentMachineName RBE:1.1 
--rtdose True 
--hideRangeShifter True 
--DoseBoxLikeCT True
#optional if wanting to compare layer by layer:
--DoseControlPoints [((BeamNumber_i,ControlPoint_i))] (using an external code that prints print(f'({beam_number},{cp_id}),') ) 
#optional if wanting to compare spot by spot:
--DoseBeamSpots [(BeamNumber_i,ControlPoint_i,BeamSpotNumber_i), ...]
```

## Script generate_scripts.py
This code will read the DICOM files generated by writeTROTStoDICOM.py and create the necessary input files for topas (compiled with the extension dicom-interface RTION)
Inside the OutputPath you choose, the following setup will be created:
OUTPUT PATH
HUtoMaterialSchneider.txt
|------Patient Folder
       |----Beam Folder
             |----run_beam.txt
             |----run_CP#.txt

>              |----run_CP#_SP#.txt

Author: ```suggestion
             |----run_CP#.txt  (if args.CPFiles == True)
             |----run_CP#_SP#.txt  (if args.SPFiles == True)
```

>              

Reviewer: If one passed the  args CPfiles spfiles

All of the previous codes are copied from [topasmc/dicom-interface](https://github.com/topasmc/dicom-interface/tree/master/rti/topas/tutorial)

### Simulations 
In order to simulate the full beamlet Range, use run_beam.txt, if you want just one control point (energy layer), choose the correct run_CP#.txt, lastly, if you want to run just one scan spot from the energy layer, run_CP#_SP#.txt                             

## Script CalibrationMU2NP.py
In this code, we will read files generated by topas for different control points (which correspond to energy layers) and on the other hand the rtdose generated for different CP

### Simulations
In order to run this script:
1) The files generated with writeTROTStoDICOM.py must have the following arguments:
```python
          --MU2NPcalibrationFile = ""
```
2) The files generated with generate_scripts.py must have the following arguments:
```python

>          --CPFiles=TRUE

Reviewer: True

>          --CPFiles=TRUE

Author: ```suggestion
         --CPFiles=True
```

         --ParticlesperHistory < 1
```
3) After this calibration is fully performed, a txt file is generated containing ernergy/calibration factor dependence. 

> 4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and stablish the following setup:

Reviewer: Stablish cambiar to add following flag to generate-scripts call

> 4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and stablish the following setup:

Author: ```suggestion
4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and add the followings flag to generate-scripts call:
```

```python
          --MU2NPcalibrationFile = "calibration_file.txt"
```

> 5) After this, the final generate_scripts.py can be rerun using a realistinc number of PPH

Reviewer: Typo realistic

Reviewer: Eg PPH 1e5

> 5) After this, the final generate_scripts.py can be rerun using a realistinc number of PPH

Author: ```suggestion
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5)
```

> 5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5)

Reviewer: Before, you need to rename all rtplanfiles to rtplan_NP.dcm or sth like this?

> 5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5)

Author: ```suggestion
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and s:Rt/RTION/PlanFile = ".../rtplan_NP.dcm"

```

> IMPORTANT: If you later want to compare the TOPAS DVH with the RTDose DVH, the final Dw.bin dose distribution must be scaled by the corresponding PPH value.

Reviewer: Could you run beam with PPH 100000 and compare DVHs wrt TROTS for a single beam? Thks

### Simulations 
In order to simulate the full beamlet Range, use run_beam.txt, if you want just one control point (energy layer), choose the correct run_CP#.txt, lastly, if you want to run just one scan spot from the energy layer, run_CP#_SP#.txt                             

## Script CalibrationMU2NP.py
In this code, we will read files generated by topas for different control points (which correspond to energy layers) and on the other hand the rtdose generated for different CP

### Simulations
In order to run this script:
1) The files generated with writeTROTStoDICOM.py must have the following arguments:
```python
          --MU2NPcalibrationFile = ""
```
2) The files generated with generate_scripts.py must have the following arguments:
```python
         --CPFiles=True
         --ParticlesperHistory < 1
```
3) After this calibration is fully performed, a txt file is generated containing ernergy/calibration factor dependence. 
4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and add the followings flag to generate-scripts call:
```python
          --MU2NPcalibrationFile = "calibration_file.txt"
```

> 5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and s:Rt/RTION/PlanFile = ".../rtplan_NP.dcm"

Reviewer: should we instead set a flag in generate_scripts --NP that automatically rewrites all scripts with rtplan_NP.dcm ?

That we should just say rerun generate_scripts with flag PPH xx and flat NP true or so

> 5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and s:Rt/RTION/PlanFile = ".../rtplan_NP.dcm"

Author: ```suggestion
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and set argument:
```python
--NP=True
```
```

OUTPUT PATH
HUtoMaterialSchneider.txt
|------Patient Folder
       |----Beam Folder
             |----run_beam.txt
             |----run_CP#.txt  (if args.CPFiles == True)
             |----run_CP#_SP#.txt  (if args.SPFiles == True)
             
All of the previous codes are copied from [topasmc/dicom-interface](https://github.com/topasmc/dicom-interface/tree/master/rti/topas/tutorial)

### Simulations 
In order to simulate the full beamlet Range, use run_beam.txt, if you want just one control point (energy layer), choose the correct run_CP#.txt, lastly, if you want to run just one scan spot from the energy layer, run_CP#_SP#.txt                             

## Script CalibrationMU2NP.py
In this code, we will read files generated by topas for different control points (which correspond to energy layers) and on the other hand the rtdose generated for different CP

### Simulations
In order to run this script:
1) The files generated with writeTROTStoDICOM.py must have the following arguments:
```python
          --MU2NPcalibrationFile = ""
```
2) The files generated with generate_scripts.py must have the following arguments:
```python
         --CPFiles=True

>          --ParticlesperHistory < 1

Author: ```suggestion
         --ParticlesperHistory < 1
         --NP=False
```

```
3) After this calibration is fully performed, a txt file is generated containing ernergy/calibration factor dependence. 
4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and add the followings flag to generate-scripts call:
```python
          --MU2NPcalibrationFile = "calibration_file.txt"
```
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and set argument:
```python

> --NP=True

Author: ```suggestion
--NP=True

```
```

## Script CalibrationMU2NP.py
In this code, we will read files generated by topas for different control points (which correspond to energy layers) and on the other hand the rtdose generated for different CP

### Simulations
In order to run this script:
1) The files generated with writeTROTStoDICOM.py must have the following arguments:
```python
          --MU2NPcalibrationFile = ""
```
2) The files generated with generate_scripts.py must have the following arguments:
```python
         --CPFiles=True
         --ParticlesperHistory < 1
         --NP=False
```
3) After this calibration is fully performed, a txt file is generated containing ernergy/calibration factor dependence. 
4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and add the followings flag to generate-scripts call:
```python
          --MU2NPcalibrationFile = "calibration_file.txt"
```
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and set argument:
```python

>            --NP=True

Reviewer: ```suggestion
           --CPFiles=False (no longer strictly needed, per-layer are only needed for the MU2NP calibration)
           --NP=True
           --PPH=1e5 (for example)
```

### Simulations 
In order to simulate the full beamlet Range, use run_beam.txt, if you want just one control point (energy layer), choose the correct run_CP#.txt, lastly, if you want to run just one scan spot from the energy layer, run_CP#_SP#.txt                             

## Script CalibrationMU2NP.py
In this code, we will read files generated by topas for different control points (which correspond to energy layers) and on the other hand the rtdose generated for different CP

### Simulations
In order to run this script:
1) The files generated with writeTROTStoDICOM.py must have the following arguments:
```python
          --MU2NPcalibrationFile = ""
```
2) The files generated with generate_scripts.py must have the following arguments:
```python
         --CPFiles=True
         --ParticlesperHistory < 1
         --NP=False
```
3) After this calibration is fully performed, a txt file is generated containing ernergy/calibration factor dependence. 
4) In order to obtain the DICOMs files with NP units, you have to return to writeTROTStoDICOM and add the followings flag to generate-scripts call:
```python
          --MU2NPcalibrationFile = "calibration_file.txt"
```

> 5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and set argument:

Reviewer: ```suggestion
5) After this, the final generate_scripts.py can be rerun using a realistic number of PPH (e.g. 1e5) and set arguments:
```
